networking: add server/client application

davidkostov opened the pull request:
Added a server/client application in which the client echoes a
message sent to it from the server.

mgenov commented on src/main/java/com/clouway/networking/serverclient/Log.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * @author David Kostov (david.kostov.cw@gmail.com)
+ */
+public class Log {
```
You can create an abstraction of this log: Display which to have implementation SwingDisplay or ConsoleDisplay. 

Coupling of direct implementations is not a good idea.

mgenov commented on src/main/java/com/clouway/networking/serverclient/TestClient.java:
```diff
@@ -0,0 +1,53 @@
+
+    public void startUp(String hostName)throws Exception{
+        socket = new Socket(hostName, port);
+        System.out.println("Client has booted up!");
```
```java
display.displayMessage("Client has booted up")
```

mgenov commented on src/main/java/com/clouway/networking/serverclient/TestClient.java:
```diff
@@ -0,0 +1,53 @@
+    public void terminate(){
+        try {
+            socket.close();
+            System.out.println("Client disconnected!");
```
display.displayMessage("Client got disconnected")

mgenov commented on src/main/java/com/clouway/networking/serverclient/TestClient.java:
```diff
@@ -0,0 +1,53 @@
+        ){
+            String message = in.readLine();
+            System.out.println("Client ~ "+message);
+            log.append("Client(echoes back) ~ "+message+"\n");
```
display.displayMessage(String.format("Client(echoes back) ~ %s\n", message))

mgenov commented on src/main/java/com/clouway/networking/serverclient/TestClient.java:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    public Socket getSocket(){
```
Doesn't look reasonable client to expose it's socket as it should be used only internally.

mgenov commented on src/test/java/com/clouway/networking/serverclient/ServerClientTask1Test.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class ServerClientTask1Test {
+
+    private Log log = new Log();
```
Instead of real implementation you can use fake one here as you are making "integrated test", e.g 2 components working together.

davidkostov commented:
Tweaked interface Log, added a Display class that implements it and extended Display into WindowDisplay(GUI log) and ConsoleDisplay. Display is abstract and holds all the log information in an array while the child classes simply display the information. Tweaked the other classes accordingly to accommodate the changes.

mgenov commented on src/main/java/com/clouway/networking/serverclient/Main.java:
```diff
@@ -0,0 +1,24 @@
+        final int port = 7777;
+        ConsoleDisplay cDisplay = new ConsoleDisplay();
+        WindowDisplay wDisplay = new WindowDisplay();
+        TestClient client = new TestClient(cDisplay, port);
```
TestClient and TestServer should be used only in tests and not in the production code. 

If client & server are developed in the same project/module then 2 separate mains should be created: ClientApp and ServerApp. The First one instantiates client which tries to connect to remote server whether the Server starts and waits for incoming connections.

mgenov commented on src/test/java/com/clouway/networking/serverclient/ServerClientTask1Test.java:
```diff
@@ -0,0 +1,48 @@
+
+        Socket socket = new Socket("localhost", port);
+        BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+        System.out.println(String.format("Client *echoes back* ~ "+in.readLine()+"\n"));
```
No assertion in this test unless you want to verify that just connections are accepted by the server but it's not clear from the name of the method.

davidkostov commented:
Declared log with List. Renamed TestServer/Client to CustomServer/Client so it makes more sense for them to be used outside of testing. Split the main in 2 parts ClientApp and ServerApp. Renamed the happPath test to simpleConnectionTest so it makes more sense.

mgenov commented on src/main/java/com/clouway/networking/serverclient/Display.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @author David Kostov (david.kostov.cw@gmail.com)
+ */
+public abstract class Display implements Log {
```
Doesn't sound reasonable Display to be and Log. The role of the Display may be to render some part of the log but it shouldn't be a log.

mgenov commented on src/main/java/com/clouway/networking/serverclient/ClientApp.java:
```diff
@@ -0,0 +1,19 @@
+        ConsoleDisplay cDisplay = new ConsoleDisplay();
+        CustomClient client = new CustomClient(cDisplay, port);
+        try {
+            client.startUp("localhost");
```
This doesn't read good. I read it like: client start up on localhost where in this case the client have to connect to "localhost:7777"  instead. 

```java
try {
  client.connect("localhost", 7777) 
  client.sendMessage("Hello !") 
} finally {
  client.disconnect() 
}
```

mgenov commented on src/test/java/com/clouway/networking/serverclient/ServerClientTask1Test.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @author David Kostov (david.kostov.cw@gmail.com)
+ */
+public class ServerClientTask1Test {
```
Name is misleading and it's not clear which of the target objects is tested. The Server or the Client. 

ServerTest 
ClientTest 

For such acceptance tests you can use either ServerAcceptanceTest or ClientAcceptanceTest or e2e naming convention such as: ServerE2ETest or ClientE2ETest

e2e = indicates that it test this component from end to end

davidkostov commented:
Removed interface Log and renamed Display to Log because the only purpose for Display in the first place was to act as a Log. Refactored the rest of the classes accordingly. Renamed client startup to connectTo also it now requires port as well as host name. Also moved the client termination in finally block.
About the tests, since there are two tests and in one of them server acceptance is tested with a random socket, while in the other the relationship between the server and the dedicated client(e.g testing both server and client simultaneously), so I wasn't sure how to name it. But since the primary behavior tested is that of the server I renamed it to Task1ServerTest instead.
